File: src/main/launchers/epic.ts
import { existsSync, readdirSync, readFileSync } from 'fs'
import { join } from 'path'
import { shell } from 'electron'
import type { GameDef, LauncherDef } from './types'

const MANIFEST_DIR = join(
    process.env['PROGRAMDATA'] ?? 'C:\\ProgramData',
    'Epic',
    'EpicGamesLauncher',
    'Data',
    'Manifests'
)

interface EpicManifest {
    DisplayName: string
    InstallLocation: string
    AppName: string
}

function readManifest(game: GameDef): EpicManifest | null {
    const epicDef = game.launchers.epic
    if (!epicDef) return null
    if (!existsSync(MANIFEST_DIR)) {
        console.warn('Epic manifest directory not found:', MANIFEST_DIR)
        return null
    }

    let files: string[]
    try {
        files = (readdirSync(MANIFEST_DIR) as string[]).filter((f) => f.endsWith('.item'))
    } catch {
        return null
    }

    for (const file of files) {
        try {
            const content = readFileSync(join(MANIFEST_DIR, file), 'utf8')
            if (!content.trim()) continue
            const manifest = JSON.parse(content) as EpicManifest
            if (manifest.DisplayName === epicDef.displayName) return manifest
        } catch (e) {
            console.warn(`Failed to parse Epic manifest ${file}:`, e)
        }
    }
    return null
}

export const EpicLauncher: LauncherDef = {
    id: 'epic',
    name: 'Epic Games',

    isInstalled(): boolean {
        return existsSync(MANIFEST_DIR)
    },

    findGame(game: GameDef): string | null {
        return readManifest(game)?.InstallLocation ?? null
    },

    identifyPath(gamePath: string): boolean {
        return existsSync(join(gamePath, '.egstore'))
    },

    launch(game: GameDef, _gamePath: string, _opts?: string): void {
        const manifest = readManifest(game)
        if (!manifest) return
        shell.openExternal(
            `com.epicgames.launcher://apps/${manifest.AppName}?action=launch&silent=true`
        )
    },
}

File: src-tauri/src/commands/launchers/epic.rs
use crate::commands::launchers::types::{GameDef, Launcher};
use std::fs;
use std::path::{Path, PathBuf};

pub struct Epic;

impl Launcher for Epic {
    fn id(&self) -> &'static str {
        "epic"
    }

    fn is_installed(&self) -> bool {
        epic_manifest_dir().exists()
    }

    fn find_game(&self, game: &GameDef) -> Option<String> {
        let def = game.epic.as_ref()?;
        epic_find_game(def.display_name).map(|(loc, _)| loc)
    }

    fn identify_path(&self, game_path: &str) -> bool {
        Path::new(game_path).join(".egstore").exists()
    }

    fn launch(&self, game: &GameDef, _game_path: &str, _opts: Option<&str>) {
        let Some(def) = game.epic.as_ref() else { return };
        if let Some((_, app_name)) = epic_find_game(def.display_name) {
            super::open_url(&format!(
                "com.epicgames.launcher://apps/{}?action=launch&silent=true",
                app_name
            ));
        }
    }
}

fn epic_manifest_dir() -> PathBuf {
    let prog_data = std::env::var("PROGRAMDATA").unwrap_or_else(|_| "C:\\ProgramData".to_string());
    PathBuf::from(prog_data)
        .join("Epic")
        .join("EpicGamesLauncher")
        .join("Data")
        .join("Manifests")
}

fn epic_find_game(display_name: &str) -> Option<(String, String)> {
    let manifest_dir = epic_manifest_dir();
    if !manifest_dir.exists() { return None; }
    for entry in fs::read_dir(&manifest_dir).ok()?.flatten() {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("item") { continue; }
        let content = fs::read_to_string(&path).unwrap_or_default();
        if content.trim().is_empty() { continue; }
        if let Ok(v) = serde_json::from_str::<serde_json::Value>(&content) {
            if v["DisplayName"].as_str() == Some(display_name) {
                let location = v["InstallLocation"].as_str()?.to_string();
                let app_name = v["AppName"].as_str().unwrap_or("").to_string();
                return Some((location, app_name));
            }
        }
    }
    None
}
